### PR TITLE
Dlg view

### PR DESCRIPTION
**Pull request by Lunalo**

### CODE REVIEW

**Comment by Lunalo** [comment]
This is ready for review @africanmathsinitiative/developers

File: instat/dlgView.vb [hunk 1]
```diff
@@ -13,46 +13,78 @@
...
+            ucrNudNumberRows.Value = ucrNudNumberRows.Maximum
         End If
     End Sub
+
     Private Sub InitialiseDialog()
+        ucrBase.iHelpTopicID = 32
+        ucrPnlDisplayWindow.AddRadioButton(rdoDispOutputWindow)
+        ucrPnlDisplayWindow.AddRadioButton(rdoDispSepOutputWindow)
+        ucrPnlDisplayFrom.AddRadioButton(rdoBottom)
+        ucrPnlDisplayFrom.AddRadioButton(rdoTop)
```
Comment (lilyclements) [inline instat/dlgView.vb]: Neither Bottom or Top are default checked

File: instat/dlgView.vb [hunk 2]
```diff
@@ -13,46 +13,78 @@
...
     End Sub
 
     Private Sub SetDefaults()
-        NumberOfRows()
+        Dim clsDefaultFunction As New RFunction
         ucrSelectorForView.Reset()
         ucrSelectorForView.Focus()
-        rdoTop.Checked = True
         rdoDispSepOutputWindow.Checked = True
         ucrSpecifyRows.Checked = True
```
Comment (lilyclements) [inline instat/dlgView.vb]: Currently, nothing is default checked out of Top/Bottom. Additionally, any rdo's and ucrCks to check need to be done in initialise

File: instat/dlgView.Designer.vb [hunk 3]
```diff
@@ -183,21 +96,113 @@ Partial Class dlgView
...
         Me.ucrBase.TabIndex = 5
         '
+        'ucrPnlDisplayWindow
+        '
+        Me.ucrPnlDisplayWindow.bAddRemoveParameter = True
+        Me.ucrPnlDisplayWindow.bChangeParameterValue = True
+        Me.ucrPnlDisplayWindow.Location = New System.Drawing.Point(10, 200)
+        Me.ucrPnlDisplayWindow.Name = "ucrPnlDisplayWindow"
+        Me.ucrPnlDisplayWindow.Size = New System.Drawing.Size(170, 46)
+        Me.ucrPnlDisplayWindow.TabIndex = 6
```
Comment (lilyclements) [inline instat/dlgView.Designer.vb]: Tab order needs to be sorted

File: instat/dlgView.vb [hunk 4]
```diff
@@ -13,46 +13,78 @@
...
+        ucrPnlDisplayWindow.AddToLinkedControls(ucrSpecifyRows, {rdoDispOutputWindow}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
+
+        ucrSpecifyRows.AddToLinkedControls(ucrPnlDisplayFrom, {True}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
+        ucrPnlDisplayFrom.SetLinkedDisplayControl(lblDisplayFrom)
+
         ucrSpecifyRows.SetText("Specify Rows")
+        ucrSpecifyRows.AddToLinkedControls(ucrNudNumberRows, {True}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
+        ucrNudNumberRows.SetLinkedDisplayControl(lblNumberofRows)
+
+        ucrSelectorForView.SetParameter(New RParameter("title"))
```
Comment (lilyclements) [inline instat/dlgView.vb]: We want to number the parameters

File: instat/dlgView.Designer.vb [hunk 5]
```diff
@@ -22,157 +22,70 @@ Partial Class dlgView
...
         Me.rdoDispOutputWindow.Text = "Display in Output Window"
         Me.rdoDispOutputWindow.UseVisualStyleBackColor = True
         '
         'ucrReceiverView
         '
+        Me.ucrReceiverView.bAddRemoveParameter = True
+        Me.ucrReceiverView.bChangeParameterValue = True
         Me.ucrReceiverView.frmParent = Me
-        Me.ucrReceiverView.Location = New System.Drawing.Point(257, 48)
+        Me.ucrReceiverView.Location = New System.Drawing.Point(248, 23)
```
Comment (lilyclements) [inline instat/dlgView.Designer.vb]: This is currently too high up, we want the top of the receiver to be level with the top of the data frames list selector (i.e., height of 60)

File: instat/dlgView.Designer.vb [hunk 6]
```diff
@@ -206,27 +211,22 @@ Partial Class dlgView
...
-    Friend WithEvents grpDisplay As GroupBox
     Friend WithEvents rdoDispSepOutputWindow As RadioButton
     Friend WithEvents rdoDispOutputWindow As RadioButton
-    Friend WithEvents lblDisplayFrom As Label
+    Friend WithEvents ucrPnlDisplayWindow As UcrPanel
+    Friend WithEvents ucrNudNumberRows As ucrNud
+    Friend WithEvents rdoBottom As RadioButton
+    Friend WithEvents rdoTop As RadioButton
+    Friend WithEvents ucrPnlDisplayFrom As UcrPanel
     Friend WithEvents ucrSpecifyRows As ucrCheck
```
Comment (lilyclements) [inline instat/dlgView.Designer.vb]: We want this to be called ucrChkSpecifyRows

File: instat/dlgView.vb [hunk 7]
```diff
@@ -13,46 +13,77 @@
...
-        clsView.SetRCommand("View")
+        ucrPnlDisplayWindow.AddToLinkedControls(ucrSpecifyRows, {rdoDispOutputWindow}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
+
+        ucrSpecifyRows.AddToLinkedControls(ucrPnlDisplayFrom, {True}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
+        ucrPnlDisplayFrom.SetLinkedDisplayControl(lblDisplayFrom)
+        ucrPnlDisplayWindow.AddFunctionNamesCondition(rdoDispSepOutputWindow, "View")
+        ucrPnlDisplayFrom.AddFunctionNamesCondition(rdoTop, "head")
+        ucrPnlDisplayFrom.AddFunctionNamesCondition(rdoBottom, "tail")
         ucrSpecifyRows.SetText("Specify Rows")
+        ucrSpecifyRows.AddToLinkedControls(ucrNudNumberRows, {True}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
```
Comment (lilyclements) [inline instat/dlgView.vb]: This nud is linked nicely in with the check box, if I uncheck the box it displays X amount of rows and if I check it it displays the amount we want.

There is a small issue, however, which I think is coming from a default in a sub later on of 6. If I have this unchecked, then it always displays just six rows. We want that when it is unchecked for it to display all the rows.

File: instat/dlgView.vb [hunk 8]
```diff
@@ -13,46 +13,77 @@
...
+        ucrPnlDisplayFrom.AddRadioButton(rdoTop)
+
         ucrReceiverView.Selector = ucrSelectorForView
         ucrReceiverView.SetMeAsReceiver()
         DataFrameLength()
-        ucrBase.iHelpTopicID = 32
-        clsView.SetRCommand("View")
+        ucrPnlDisplayWindow.AddToLinkedControls(ucrSpecifyRows, {rdoDispOutputWindow}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
+
+        ucrSpecifyRows.AddToLinkedControls(ucrPnlDisplayFrom, {True}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
```
Comment (lilyclements) [inline instat/dlgView.vb]: Would we want to have ucrChkSpecifyRows.SetRDefault("TRUE") @dannyparsons ? This isn't an R Default, but I'm not sure how else we specify that when there is linking?
This probably is similarly seen with the ucrNud. We need to set a default value, but not necessarily the R Default?

File: instat/dlgView.vb [hunk 9]
```diff
@@ -13,46 +13,77 @@
...
+        ucrPnlDisplayWindow.AddFunctionNamesCondition(rdoDispSepOutputWindow, "View")
+        ucrPnlDisplayFrom.AddFunctionNamesCondition(rdoTop, "head")
+        ucrPnlDisplayFrom.AddFunctionNamesCondition(rdoBottom, "tail")
         ucrSpecifyRows.SetText("Specify Rows")
+        ucrSpecifyRows.AddToLinkedControls(ucrNudNumberRows, {True}, bNewLinkedHideIfParameterMissing:=True, bNewLinkedAddRemoveParameter:=True)
+        ucrNudNumberRows.SetLinkedDisplayControl(lblNumberofRows)
+        ucrSelectorForView.SetParameter(New RParameter("title", 1))
+        ucrSelectorForView.SetParameterIsString()
+
+        ucrNudNumberRows.SetParameter(New RParameter("n", 1))
```
Comment (lilyclements) [inline instat/dlgView.vb]: The nud and the selector have been given the same parameter order value. Was this intentional?

File: instat/dlgView.vb [hunk 10]
```diff
@@ -61,127 +92,50 @@ Public Class dlgView
...
-        SetCommands()
-        TestOKEnabled()
-    End Sub
-
-    Private Sub SetCommands()
-        If rdoDispSepOutputWindow.Checked Then
-            ucrBase.clsRsyntax.SetBaseRFunction(clsView)
-            ucrBase.clsRsyntax.AddParameter("x", clsRFunctionParameter:=ucrReceiverView.GetVariables())
-        ElseIf rdoDispOutputWindow.Checked Then
+    Private Sub SetIcallType()
```
Comment (lilyclements) [inline instat/dlgView.vb]: So, my understanding was that changing functions between heads and tails was something done in initialise. I ran it without the "Set Function" part of this sub and it didn't work. I'm just intrigued why we need to have it in both?

File: instat/dlgView.Designer.vb (hunk 3 above)
Comment (Lunalo) [inline instat/dlgView.Designer.vb]: This has now been sorted

File: instat/dlgView.Designer.vb (hunk 5 above)
Comment (Lunalo) [inline instat/dlgView.Designer.vb]: Thanks. it has now been sorted

File: instat/dlgView.Designer.vb (hunk 6 above)
Comment (Lunalo) [inline instat/dlgView.Designer.vb]: haha, that eye. I have fixed it

File: instat/dlgView.vb (hunk 7 above)
Comment (Lunalo) [inline instat/dlgView.vb]: This is a bit technical. I tried my best and agreed wit Danny to look at it.

File: instat/dlgView.vb (hunk 8 above)
Comment (Lunalo) [inline instat/dlgView.vb]: ucrChkSpecifyRows this just controling the hiding and unhding parameters. It has no parameter that is passing the value "TRUE"

File: instat/dlgView.vb (hunk 10 above)
Comment (Lunalo) [inline instat/dlgView.vb]: We need to do this and pass the sub in Value Changed of respective control to trigger the updating of r code

File: instat/dlgView.vb (hunk 9 above)
Comment (Lunalo) [inline instat/dlgView.vb]: The title parameter is for View function while n is for both head and tail. The parameters are being passed in different function.

I did it after carefully thinking of how it was functioning. Thanks

**Pull request merged**